src/Persistence_matrix/include/gudhi/Persistence_matrix/columns/heap_column.h

```diff
@@ -0,0 +1,1071 @@
+inline Heap_column<Master_matrix>& Heap_column<Master_matrix>::multiply_source_and_add(
+    const Cell_range& column, const Field_element_type& val) 
+{
+  static_assert((!Master_matrix::isNonBasic || std::is_same_v<Cell_range, Heap_column>),
+                "For boundary columns, the range has to be a column of same type to help ensure the validity of the "
+                "base element.");  // could be removed, if we give the responsability to the user.
+  static_assert((!Master_matrix::isNonBasic || Master_matrix::Option_list::is_of_boundary_type),
+                "For chain columns, the given column cannot be constant.");
+
+  if constexpr (Master_matrix::Option_list::is_z2) {
+    if (val) {
+      _add(column);
+    }
+  } else {
+    _multiply_source_and_add(column, val);
+  }
+
+  return *this;
+}
+
+template <class Master_matrix>
+inline Heap_column<Master_matrix>& Heap_column<Master_matrix>::multiply_source_and_add(
+    Heap_column& column, const Field_element_type& val) 
+{
+  if constexpr (Master_matrix::isNonBasic && !Master_matrix::Option_list::is_of_boundary_type) {
+    // assumes that the addition never zeros out this column.
+    if constexpr (Master_matrix::Option_list::is_z2) {
+      if (val) {
+        if (_add(column)) {
+          chain_opt::swap_pivots(column);
+          dim_opt::swap_dimension(column);
+        }
+      }
+    } else {
+      if (_multiply_source_and_add(column, val)) {
+        chain_opt::swap_pivots(column);
+        dim_opt::swap_dimension(column);
+      }
+    }
+  } else {
+    if constexpr (Master_matrix::Option_list::is_z2) {
+      if (val) {
+        _add(column);
+      }
+    } else {
+      _multiply_source_and_add(column, val);
+    }
+  }
+
+  return *this;
+}
+
+template <class Master_matrix>
+inline Heap_column<Master_matrix>& Heap_column<Master_matrix>::operator=(const Heap_column& other) 
+{
+  static_assert(!Master_matrix::Option_list::has_row_access, "= assignement not enabled with row access option.");
+
+  dim_opt::operator=(other);
+  chain_opt::operator=(other);
+
+  while (column_.size() > other.column_.size()) {
+    if (column_.back() != nullptr) cellPool_->destroy(column_.back());
+    column_.pop_back();
+  }
+
+  column_.resize(other.column_.size(), nullptr);
+  index i = 0;
+  for (const Cell* cell : other.column_) {
+    if (column_[i] != nullptr) {
+      cellPool_->destroy(column_[i]);
+    }
+    column_[i] = other.cellPool_->construct(cell->get_row_index());
+    if constexpr (!Master_matrix::Option_list::is_z2) {
+      column_[i]->set_element(cell->get_element());
+    }
+    ++i;
+  }
+  insertsSinceLastPrune_ = other.insertsSinceLastPrune_;
+  operators_ = other.operators_;
+  cellPool_ = other.cellPool_;
+
+  return *this;
+}
+
+template <class Master_matrix>
+inline std::size_t Heap_column<Master_matrix>::compute_hash_value()
+{
+  _prune();
+  //can't use Gudhi::persistence_matrix::hash_column because the same heap can ordered differently
```

Comment:
If `_prune` actually sorts, it looks like you could rely on the order here?

Reply:
Right, I completely missed that!